tests/framework/e2e/cluster.go

```diff
@@ -306,6 +310,16 @@ func NewEtcdProcessCluster(ctx context.Context, t testing.TB, cfg *EtcdProcessCl
 	return StartEtcdProcessCluster(ctx, epc, cfg)
 }
 
+func NewEtcdProcessClusterWithOptions(ctx context.Context, t testing.TB, opts ...EPClusterOption) (*EtcdProcessCluster, error) {
+	cfg := NewConfig(opts...)
+	epc, err := InitEtcdProcessCluster(t, cfg)
+	if err != nil {
+		return nil, err
+	}
+
+	return StartEtcdProcessCluster(ctx, epc, cfg)
+}
+
```

Comment:
For smooth transition, `NewEtcdProcessClusterWithOptions` was created.
After all initialisations are replaced, `NewEtcdProcessClusterWithOptions` will be renamed to `NewEtcdProcessCluster`, and the original `NewEtcdProcessCluster` will be deleted.

Reply:
sg

Reply:
I'd suggest to enhance the existing function `NewEtcdProcessCluster` instead of adding a new one. See code example below. Note that it has no any impact on existing test cases. 

```
func NewEtcdProcessCluster(ctx context.Context, t testing.TB, cfg *EtcdProcessClusterConfig, opts ...EPClusterOption) (*EtcdProcessCluster, error) {
	if cfg == nil {
		cfg = NewConfig(opts...)
	} else {
		for _, opt := range opts {
			opt(cfg)
		}
	}
	epc, err := InitEtcdProcessCluster(t, cfg)
	if err != nil {
		return nil, err
	}

	return StartEtcdProcessCluster(ctx, epc, cfg)
}
```

Reply:
It looks much better👍 I'm gonna change code to this.

After the transition period, I envision it eventually going to be this:
```go
func NewEtcdProcessCluster(ctx context.Context, t testing.TB, opts ...EPClusterOption) (*EtcdProcessCluster, error) {
 	cfg := NewConfig(opts...)
 	epc, err := InitEtcdProcessCluster(t, cfg)
 	if err != nil {
 		return nil, err
 	}
 	return StartEtcdProcessCluster(ctx, epc, cfg)
 }
```

And there will be 2 ways to call `NewEtcdProcessCluster`:
```go
epc, err := NewEtcdProcessCluster(ctx, t,
 	e2e.WithKeepDataDir(true),
 	e2e.WithCorruptCheckTime(time.Second),
)
```
```go
epc, err := NewEtcdProcessCluster(ctx, t, e2e.WithConfig(cfg))
```
What do you think?

Reply:
Basically makes sense to me. 



> And there will be 2 ways to call `NewEtcdProcessCluster`

There is only one way. 

```
epc, err := NewEtcdProcessCluster(ctx, t, e2e.WithConfig(cfg), e2e.WithKeepDataDir(true))
```

Ideally, if we follow this approach, then all fields in `EtcdProcessClusterConfig` shouldn't be exported, but of course it can be addressed in a separate PR.

Reply:
> Ideally, if we follow this approach, then all fields in `EtcdProcessClusterConfig` shouldn't be exported, but of course it can be addressed in a separate PR.

This comment might not be correct. As long as we keep the `e2e.WithConfig`, then the fields are still needed to be exported.

Reply:
> This comment might not be correct. As long as we keep the `e2e.WithConfig`, then the fields are still needed to be exported.

@ahrtr  From my perspective, keeping `e2e.WithConfig` doesn't require fields in `EtcdProcessClusterConfig` to be exported. Did I overlook something?

```
cfg :=NewConfig(
	WithClientTLS(ClientTLS),
	WithIsPeerTLS(true),
)

cluster := NewEtcdProcessCluster(ctx, t, WithConfig(cfg))
```

Reply:
Keeping the fields exported means keeping the flexibility to specify the values for fields directly, and then pass it to `NewEtcdProcessCluster` using `WithConfig(cfg)`